text/0000-inline-asm.md

```diff
@@ -0,0 +1,980 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Inline assembler is implemented as an unsafe macro `asm!()`.
+The first argument to this macro is a template string literal used to build the final assembly.
+The following arguments specify input and output operands.
+When required, options are specified as the final argument.
+
+The following ABNF specifies the general syntax:
+
+```
+dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"
+reg_spec := <register class> / "<explicit register>"
+operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"
+reg_operand := dir_spec "(" reg_spec ")" operand_expr
+operand := reg_operand / "const" const_expr / "sym" path
+option := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn"
+options := "options(" option *["," option] ")"
+asm := "asm!(" format_string *("," [ident "="] operand) ["," options] ")"
+```
+
+[format-syntax]: https://doc.rust-lang.org/std/fmt/#syntax
+
+## Template string
+
+The assembler template uses the same syntax as [format strings][format-syntax] (i.e. placeholders are specified by curly braces). The corresponding arguments are accessed in order, by index, or by name. However, implicit named arguments (introduced by [RFC #2795][rfc-2795]) are not supported.
+
+As with format strings, named arguments must appear after positional arguments. Explicit register operands must appear at the end of the operand list, after any named arguments if any. Explicit register operands cannot be used by placeholders in the template string. All other operands must appear at least once in the template string, otherwise a compiler error is generated.
+
+The assembly code syntax used is that of the GNU assembler (GAS). The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.
+
+This RFC only specifies how operands are substituted into the template string. Actual interpretation of the final asm string is left to the assembler.
+
+However there is one restriction on the asm string: any assembler state (e.g. the current section which can be changed with `.section`) must be restored to its original value at the end of the asm string.
+
+[rfc-2795]: https://github.com/rust-lang/rfcs/pull/2795
+
+## Operand type
+
+Several types of operands are supported:
+
+* `in(<reg>) <expr>`
+  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
+  - The allocated register will contain the value of `<expr>` at the start of the asm code.
+  - The allocated register must contain the same value at the end of the asm code (except if a `lateout` is allocated to the same register).
+* `out(<reg>) <expr>`
+  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
+  - The allocated register will contain an unknown value at the start of the asm code.
```

Comment:
I think elsewhere you call this "undefined value"? Is that the same as "unknown value"?
Both terminologies should be precisely defined. (I hope to avoid the situation in C where "indeterminate value" is full of ambiguities.) If I saw it correctly, you are only using "undefined value" in text that is written from the view of assembly, never on the level of the Abstract Machine, so I suspect the definition is "at the *beginning* of the execution of the asm block, non-deterministically pick any possible value this register can have in the target architecture" -- in particular, once the value is chosen, it is a "normal value", not something "unstable" that can change every time we look at it.